### ISSUE TITLE
[proposal] default to MELPA Stable

### ISSUE THREAD

**Issue by dunn** [milkypostman/melpa#3306]
I don't know if there are stats, but I'm guessing the vast majority of users are on MELPA Unstable rather than Stable.  I'd like to suggest that there should be some effort made to get more users on the Stable archive.

It seems like a logical step to take as a result of [the maturation of the ecosystem](http://www.emacswiki.org/emacs/2014-10-30) that Alex Schroeder cited when he told people to stop uploading code to EmacsWiki.  One of the things he thought users are now (or should be) demanding is _proper releases_.  Promoting proper—tagged, versioned—releases is good for users and developers; users get more stability (and maybe eventually the power to install specific versions) and developers are encouraged to think more about backwards compatibility and in general more \~user conscious developement\~, especially if we encourage semver.

But! I'm aware that Stable is underdeveloped and not actually used by the maintainers themselves (https://github.com/milkypostman/melpa/issues/3023#issuecomment-131473533, https://github.com/milkypostman/melpa/issues/2565#issuecomment-77275372), meaning that switching to Stable might provide a _worse_ user experience if it happened today.  But as things are now, there's very little motivation for contributors to improve Stable, or for package developers to tag stable releases (#1786).

So I think it'd be a really good idea to switch the default archive to Stable sometime in 2016.  Emacs 25 will be able to detect uninstallable packages (https://github.com/milkypostman/melpa/issues/1955#issuecomment-102679358), and once Stable supports the full range of fetchers (#2075) that would remove some of the major technical obstacles (unless there are others I'm not aware of).  I might take a stab at improving version detection (#2553) this weekend.  Of course the task would remain of convincing developers to get on board with regular versioned releases, but MELPA probably has enough clout to help drive these sorts of changes.

(Full disclosure: I'm the maintainer of the [Homebrew Emacs tap](https://github.com/Homebrew/homebrew-emacs) which does _not_ have that sort of clout; Homebrew is not very good at handling `HEAD` builds so this would help me there.)

**Comment by vyp** [comment]
You mean like make "melpa" actually what "melpa stable" is now, and then make what "melpa" is now something like "melpa unstable"?

**Comment by dunn** [comment]
Yeah, exactly.

**Comment by purcell** [comment]
Thanks for taking the time to write this up. Ultimately this is a chicken and egg situation. There has to be some investment of time and effort on the part of the community if Stable is ever to be viable, but nobody has yet stepped up, even to simply implement fetchers for the SCMs we don't yet support. I think that what you're proposing is that "we" instead come up with a plan to push everyone towards stable packages.

My personal suspicion is that while people like the idea of stable packages, and have a feeling that their experience of using stable packages would be better somehow, the vast majority of people are using the snapshot packages without problems, and have no particular interest in changing.

It's not particularly clear to me which problems "stable" packages would solve in the context of Emacs. Snapshot packages which are broken accidentally will typically get fixed and rebuilt quickly, and new stable versions of packages are at least as likely to break a user's config when the user upgrades. Since `package.el` doesn't support version pinning or installation of exact versions, the only way to avoid breakage in the case of either snapshot _or_ stable packages is to avoid upgrading currently-working packages anyway.

So right now I can't imagine making a big push to move people over to MELPA Stable -- I couldn't even tell any user what the benefits might be!

**Comment by dunn** [comment]
> I couldn't even tell any user what the benefits might be!

Haha, yeah.  I'm surprised there haven't been many major package breakages, but I guess there just aren't enough large/complex applications that it's a real concern yet?  Maybe the chicken-egg situation will resolve itself without artificial pressure, if we get to the point where there are those sorts of applications, and demand for a stable distribution channel arises naturally.

**Comment by rnkn** [comment]
I'd be quite happy to see the site with two big buttons: STABLE and BETA, the former linking to https://stable.melpa.org, the latter to https://beta.melpa.org.

The packages would then be served at https://stable.melpa.org/packages/ and https://beta.melpa.org/packages/ respectively.

Shortly after that the original https://melpa.org/packages/ would go dark.

**Comment by purcell** [comment]
@rnkn Does that proposal address any the concerns above? I feel like it's just a renaming which wouldn't make any positive difference, so perhaps you haven't quite explained why you'd be so happy to see those big buttons?

**Comment by rnkn** [comment]
Are you saying my happiness is *not* the most important thing here?

Yeah I think it would do two things:

- Address the perception, both with developers and users, that MELPA is primary and MELPA-stable is secondary. It's pretty common that package authors just link to https://melpa.org so if I'm a newbie it's unlikely I even know that MELPA-stable exists.
- If the current package archive is scheduled to go offline one day, then that day (or some point after) everyone chooses between two perceived equally-weighted options, which would reinforce the previous.

So, if I'm new to Emacs and I'm linked to or told that packages are "available on MELPA" then my first impression is now quite different. I'm probably more likely to go for the stable package and/or ask developers why their package is "still only in beta."

And now Developers who were previously content without MELPA-stable because MELPA was "the main one" will likely react differently, just because their hard work is called "beta".

**Comment by Silex** [comment]
This `stable/beta` renaming idea is actually quite logical and is pretty easy to do.

The phrasing would change from:

``` markdown
# MELPA

MELPA is a growing collection of package.el-compatible Emacs Lisp packages built automatically on our server from the upstream source code using simple recipes. (Think of it as a server-side version of el-get, or even Homebrew.)

....

## MELPA Stable

By default, MELPA provides the very latest revisions of packages. If you prefer to only receive updates for tagged releases, use MELPA Stable instead.
```

To something like:

``` markdown
# MELPA

MELPA is a growing collection of package.el-compatible Emacs Lisp packages built automatically on our server from the upstream source code using simple recipes. (Think of it as a server-side version of el-get, or even Homebrew.)

MELPA comes in two versions: 

- **stable**: provides the tagged releases of packages only.
- **beta**: provides the very latest revisions of packages.
```

**Comment by dunn** [comment]
I think 'beta' would be the wrong name for the bleeding-edge channel; 'beta' usually refers to releases which are superseded by a stable release.  Something like 'devel',  or just 'edge' might be better.

**Comment by rnkn** [comment]
I like "beta" mostly because anyone who has watched CSI Cyber has a vague idea of what beta software is. Emacs for the masses!

**Comment by purcell** [comment]
Thanks for the considered feedback - much appreciated.

> Address the perception, both with developers and users, that MELPA is primary and MELPA-stable is secondary.

This is actually the crux of things: from our point of view, that's exactly the case. I wouldn't recommend MELPA Stable to anyone, to be honest, because I haven't personally found it useful.

> It's pretty common that package authors just link to https://melpa.org so if I'm a newbie it's unlikely I even know that MELPA-stable exists.

That could also mean that everyone is using MELPA, so MELPA Stable is a distraction or red herring. I wouldn't know how to fairly compare the current popularity of the two archives, but based on download history, MELPA Stable is running at 775 downloads/pkg over all time vs 9000 dl/pkg for MELPA. Packages in the latter will be upgrade more often, but even so, that's a huge discrepancy.

**Comment by Silex** [comment]
>> Address the perception, both with developers and users, that MELPA is primary and MELPA-stable is secondary.

> This is actually the crux of things: from our point of view, that's exactly the case. I wouldn't recommend MELPA Stable to anyone, to be honest, because I haven't personally found it useful.

As I implied earlier, if that's the vision then I'm fine with the current state. I'm also a user of MELPA "beta".

**Comment by rnkn** [comment]
I think Emacs has seen a steady move of UX from complicated and developer-centric, to easier and user-centric. Regardless of how things are currently, I think the future of Emacs will look like:

1. user launches Emacs
2. an asynchronous process fetches packages that can be upgraded
3. the user is told that packages can be upgraded, do you want to install/review/ignore
4. reviewing packages results in a list of "package foo version X.X is available, you have version X.Y" with a change log of version X.Y => X.X
5. upgrading happens in the background

Despite the huge discrepancy between stable and snapshot now, that will eventually go away.

I'm willing to bet 500 Emacs points on this.

My thinking is less about what is better or right and more that this will eventually happen regardless  of whether it's with package.el + MELPA or something that supplants them, so why fight the future?

**Comment by alphapapa** [comment]
I think the elephant in the room is development methodology, or the styles and preferences of individual developers.  I try to use semantic versioning-style development for my packages, tagging stable releases for *stable* while also maintaining a version in *non-stable* that's usable but might have incompatible changes without warning.  

Of course most downloads of my packages have been from *non-stable*, but I have noticed a significant uptake in downloads from *stable* (although with the recent problems with download counters, I'm not sure what to make of it).

Anyway, the issue as I see it is that, as much as some of us might like to see all software developed this way, MELPA can't force package authors to do so.  If an author doesn't want to do it semver-style, or just doesn't want to think about stability and compatibility, we can't make them.  If they want their only releases to be the current HEAD, well, it's their package, and it's up to users whether they want to use it.  I figure MELPA exists to enable users to do what they need and want to do, not to force authors into a certain style, even if using that style would be good for users.

Having said that, I think there might be room for MELPA to begin gently nudging authors in the direction of considered, stable releases.  But as Steve pointed out, it would probably mean changing the mindset toward the *stable* repo being the primary one recommended to users.  And since that's not currently the way it is, there might be something to the idea of shaking things up a little with name changes.  

I'm not sure "beta" is the most accurate term for what the current *non-stable* repo is, but maybe something else...  Having used Debian for many years, both stable releases and rolling testing/unstable ones, I wonder if it would be good to follow Debian's example and have MELPA Stable and MELPA Testing repos.  Since "unstable" can be interpreted to mean something like "crashes," maybe it wouldn't be a good choice, but "testing" might work, since it could mean "this package is not as stable as the one in the Stable repo; it's currently being tested, and should hopefully be promoted to the Stable repo soon.  Feel free to test it if you like, because we always need more testers."

**Comment by purcell** [comment]
> I think there might be room for MELPA to begin gently nudging authors in the direction of considered, stable releases.

Yeah, I think that if #2955 works out, and the MELPA version scheme changes, then there'll be a bigger visual difference on MELPA between packages which are tagged for release and those which are only ever released as snapshots.

It'd also perhaps be nice to mark packages on MELPA which are also in MELPA Stable.

I'm less keen to rename MELPA itself, but it's something to consider perhaps.

**Comment by alphapapa** [comment]
> It'd also perhaps be nice to mark packages on MELPA which are also in MELPA Stable.

I haven't upgraded yet, but with pinning in Emacs 25, would this make it easier for people to mix the two repos, preferring packages from Stable?

**Comment by purcell** [comment]
Yes, I believe so.

**Comment by Wilfred** [comment]
Is it worth providing a link to MELPA stable on the melpa.org homepage? AFAICS there's no mention of it on there currently.

**Comment by tarsius** [comment]
I've assigned this to myself as a way to keep track of it. This does not necessarily mean I will do something.

**Comment by Fuco1** [comment]
@purcell writes

> It's not particularly clear to me which problems "stable" packages would solve in the context of Emacs. 

While this is true for the "application" packages, it is a big big big issue for libraries.  I am now toying with the idea of finally releasing dash version 3 which will be backwards incompatible.  Melpa makes this task impossible.  Doing it would immediately break *every* Emacs user's setup out there for a considerable time until all the packages migrated to the new version, which could take months.

I am thinking hard about how to do the rollout and will probably start a new issue to ask for help.  But with the advances in the ecosystem and many many libraries available to Emacs as packages this is something we now need to reconsider.

**Comment by purcell** [comment]
@Fuco1 Yep, ultimately the solution is something like #2955, which addresses this unfortunate case. I agree such roll-outs are difficult currently.

**Comment by minad** [comment]
I suggest that the contradicting proposals #3306 and #6656 are both closed, given the new channels snapshots.melpa.org and releases.melpa.org (#10057).

**Comment by tarsius** [comment]
- #3306 proposes that we "default to MELPA Stable"
- #6656 proposes that we "abolish MELPA Stable"

Clearly different users want different things.

We have long ago decided, though maybe not communicated explicitly enough, that we are not going implement either proposal. Keeping things as-is is the best compromise. Users who want to use MELPA Stable can continue to do so and users who don't want to use it don't have to and don't even have to do anything to avoid doing so.

We, the MELPA maintainer, continue to recommend that users use "regular" MELPA (or now the experimental MELPA Snapshots"). Personally hope that using MELPA Stable (or rather the experimental replacement MELPA Releases) becomes feasible one day. I hope that the new releases channel increases the odds of that happening, but we have a long way to go. Many maintainers have *never* created a release, many other maintainers have stopped releasing somewhere along the way, and yet others provide conflicting release information (the tag and version header do not agree).

I touch on these topics (including why using "stable" does not lead to a more stable experience) in [Announcing two new experimental MELPA channels](https://emacsair.me/2026/07/03/melpa-channels/).
